javalib/src/main/scala/java/util/PriorityQueue.scala

```diff
@@ -274,13 +319,88 @@ class PriorityQueue[E] private (
 
         // if the node `m` is greater than the selected child, swap and recurse
         if (comp.compare(innerAtM, innerAtJ) > 0) {
-          inner(m) = innerAtJ
-          inner(j) = innerAtM
+          innerImpl.set(inner, m, innerAtJ)
+          innerImpl.set(inner, j, innerAtM)
           loop(j)
         }
       }
     }
 
     loop(m)
   }
+
+  private sealed abstract class InnerArrayImpl {
+    type Repr <: AnyRef
+
+    def make(initialCapacity: Int): Repr
+    def length(v: Repr): Int
+    /** Set the length of innerArray.
+      *
+      * In WebAssembly, freeing the reference for GC is needed
+      * when we shrink the inner array.
+      */
+    def setLength(v: Repr, newLength: Int): Unit
+    def get(v: Repr, index: Int): E
+    def set(v: Repr, index: Int, e: E): Unit
+    def push(v: Repr, e: E): Unit
+    /** Wasm only. */
+    def resized(v: Repr, minCapacity: Int): Repr
+    /** Wasm only. */
+    def capacity(v: Repr): Int
+    def copyFrom(v: Repr, from: Int): Repr
+    def clear(v: Repr): Unit
+  }
+
+  private object InnerArrayImpl {
+    object JSArrayImpl extends InnerArrayImpl {
+      type Repr = js.Array[E]
+
+      // The index 0 is not used; the root is at index 1.
+      // This is standard practice in binary heaps, to simplify arithmetics.
+      @inline def make(_initialCapacity: Int): Repr = js.Array[E](null.asInstanceOf[E])
+      @inline def length(v: Repr): Int = v.length
+      @inline def setLength(v: Repr, newLength: Int): Unit =
+        v.length = newLength
+      @inline def get(v: Repr, index: Int): E = v(index)
+      @inline def set(v: Repr, index: Int, e: E): Unit =
+        v(index) = e
+      @inline def push(v: Repr, e: E): Unit =
+        v.push(e)
+      @inline def resized(v: Repr, minCapacity: Int): Repr = v // no used
+      @inline def capacity(v: Repr): Int = 0 // no used
+      @inline def copyFrom(v: Repr, from: Int): Repr =
+        v.jsSlice(from)
+      @inline def clear(v: Repr): Unit =
+        v.length = 1
+    }
+
+    object JArrayImpl extends InnerArrayImpl {
+      type Repr = Array[AnyRef]
+
+      // The index 0 is not used; the root is at index 1.
+      // This is standard practice in binary heaps, to simplify arithmetics.
+      private var _size = 1
```

Comment:
Something went wrong during the refactoring, because this state variable was moved to the global impl object. 😅

Reply:
Oops, thanks, moved back to the `PriorityQueue`